### DISCUSSION TITLE
New `transformTestFile` option to interface with bundlers

### DISCUSSION THREAD

**Discussion by Prinzhorn** [avajs/ava#3074] [Ideas]
I'm using AVA for unit tests and it has worked great so far. Now I wanted to add integration tests, but AVA (or Node.js rather) chokes on `.sql` imports. I import `.sql` migration files using the esbuild `text` loader. This way I end up with a single bundle and faster startup time (Electron).

I'm sure there are multiple ways to work my way around this, e.g. having my tests as `*.build.test.js`, ignore them in AVA config and have an esbuild process output bundled `.test.js` files without the `build` part, which AVA will pick up. But that doesn't sound fun. What does sound fun would be something like this:

```js
export default {
  transformTestFile: async (file) => {
    let transformed = await esbuildMagic(file);
    return transformed;
  },
};
```

I assume other bundlers could use this interface in a similar fashion and transforms could be provided for ts and whatnot.

```js
import { transform } from 'my-svelte/vue/react/ts-transform';

export default {
  transformTestFile: transform
};
```

From what I can tell right now there is no unified way to do things. And I'm definitely not going to try to integrate babel to do something that esbuild is already doing for my actual code. To me this looks cleaner than any of ts-node, babel, require-hooks, --experimental-loader and whatnot.

**Comment by novemberborn** [comment]
`@ava/typescript` uses an extensible protocol for hooking into test file selection, pre-compilation and loading. It's just not generically available, which is a change we can make.

Note though that this only impacts the test files themselves, which does suit your use case. For other approaches loader hooks are more appropriate.

**Reply by Prinzhorn** [reply]
Sounds good to me. If you could abstract this into some sort of generic `@ava/transform` (which `@ava/typescript` uses/extends) I could probably use that to hook into the appropriate life cycle parts? Before polishing this you could point me to some duct taped version that I could evaluate with my setup. I'd be happy to test that and give feedback if it works or if we are missing something (e.g. I've declared some modules as external because they cannot be bundled, but that shouldn't be an issue, since the transpiled test file can still access the same `node_modules`)

**Reply by Prinzhorn** [reply]
One more thing I was curious and you can probably answer: from what I can tell `@ava/typescript` actually creates compiled files in a different folder. Would it be possible to pass `data:` or `eval: true` to each Worker? Because with esbuild I don't see a reason to write the bundled test files to the file system at all, I can use `write: false` and get the string back. Did you ever experiment with that in AVA? Would the tests even work when the worker uses a string of JavaScript? Or does that break some assumptions AVA makes?

**Reply by novemberborn** [reply]
There's already a fairly generic protocol:

* https://github.com/avajs/ava/blob/6f8e30c94626238a5b26deadac319089fa43d333/lib/provider-manager.js
* https://github.com/avajs/ava/blob/6f8e30c94626238a5b26deadac319089fa43d333/lib/globs.js#L69-L71
* https://github.com/avajs/ava/blob/6f8e30c94626238a5b26deadac319089fa43d333/lib/globs.js#L186
* https://github.com/avajs/ava/blob/6f8e30c94626238a5b26deadac319089fa43d333/lib/api.js#L241-L247
* https://github.com/avajs/ava/blob/6f8e30c94626238a5b26deadac319089fa43d333/lib/worker/base.js#L126-L139

It would only take a few tweaks to make that fully extensible (without requiring the provider to be recognized by AVA).

Regarding the loading, AVA still very much wants to require or import, but the provider can control the URL that is imported so that could be a data URL. See the `worker/base` file above.

**Reply by Prinzhorn** [reply]
Thanks a lot, I'll look into this in the future. It's not a priority right now but I'll come back to it when I figured some other stuff out.
